---
title: "BCB420 - Computational Systems Biology Assigment 2"
subtitle: "Differential Gene expression and Preliminary ORA"
author: "Joshua Efe"
output:
  html_document:
    toc: true
    toc_depth: 2
bibliography: a.bib

---
This is the Needed code to run from A1
```{r nest a1, message=FALSE, warning=FALSE, child='A1.Rmd', include=FALSE, echo=FALSE, eval=TRUE}
library(ComplexHeatmap)
library(circlize)
```


```{r}
heatmap_matrix <- NormalizedData[,2:ncol(NormalizedData)]
rownames(heatmap_matrix) <- NormalizedData$genes
colnames(heatmap_matrix) <- colnames(NormalizedData[,2:ncol(NormalizedData)])

if(min(heatmap_matrix) == 0){
    heatmap_col = colorRamp2(c( 0, max(heatmap_matrix)), c( "white", "red"))
  } else {
    heatmap_col = colorRamp2(c(min(heatmap_matrix), 0, max(heatmap_matrix)), c("blue", "white", "red"))
  }
current_heatmap <- Heatmap(as.matrix(heatmap_matrix),
                               show_row_dend = TRUE,
                               show_column_dend = TRUE, 
                               col=heatmap_col,
                               show_column_names = TRUE, 
                               show_row_names = FALSE,
                               show_heatmap_legend = TRUE
                               )
current_heatmap
```

First we are going to calculate the p values for each of the genes but first a quick example and how the data fits together with an MDS Plot 
```{r message=FALSE, warning=FALSE}
#An example of a gene with two of the different samples
iPSHPC2EGFP <- grep(colnames(NormalizedData),
                          pattern="EGFP")
iPSmicroglia2H9 <- grep(colnames(NormalizedData),
                          pattern="H9")
gene_of_interest <- which(NormalizedData$gene == "A1BG")

#_____
iPSHPC2EGFP_samples <- t(NormalizedData
                       [gene_of_interest,
                         iPSHPC2EGFP])
colnames(iPSHPC2EGFP_samples) <- c("iPSHPC2EGFP_samples")
iPSHPC2EGFP_samples

#--
iPSmicroglia2H9_samples <- t(NormalizedData
                       [gene_of_interest,
                         iPSmicroglia2H9])
colnames(iPSmicroglia2H9_samples) <- c("iPSmicroglia2H9_samples")
iPSmicroglia2H9_samples
#----
t.test(x=t(iPSHPC2EGFP_samples),y=t(iPSmicroglia2H9_samples))


```


Lets now calculate the p values of all the genes
```{r message=FALSE, warning=FALSE}

samples <- data.frame(
        lapply(colnames(NormalizedData)[2:22], 
        FUN=function(x){
          unlist(strsplit(x, split = "(([P])|([5-9]))_"))[c(1,2)]}))
colnames(samples) <- colnames(NormalizedData)[2:22]
rownames(samples) <- c("cell_type","repID") 
samples <- data.frame(t(samples))
samples


model_design <- model.matrix(~ samples$cell_type)
kable(model_design, type="html")

#create a dataMatrix
expressionMatrix <- as.matrix(NormalizedData[,2:22])
rownames(expressionMatrix) <- NormalizedData$gene
colnames(expressionMatrix) <- colnames(NormalizedData)[2:22]
minimalSet <- ExpressionSet(assayData=expressionMatrix)

#fit data to the above model
fit <- lmFit(minimalSet, model_design)

#Apply empircal Bayes to expression data
fit2 <- eBayes(fit,trend=TRUE)

topfit <- topTable(fit2, 
                   coef=ncol(model_design),
                   adjust.method = "BH",
                   number = nrow(expressionMatrix))
#merge hgnc names to topfit table
output_hits <- merge(NormalizedData[,1],
                     topfit,
                     by.y=0,by.x=1,
                     all.y=TRUE)
#sort by pvalue
output_hits <- output_hits[order(output_hits$P.Value),]
kable(output_hits[1:10,],type="html")

#how many p values less than 0.05?
length(which(output_hits$P.Value < 0.05))

#How many with corrected p values less than 0.05 (multiple hypothsis testing) #25
length(which(output_hits$adj.P.Val < 0.05))

```
From here we can see there seems to be a lot of signficantly expressed genes 11090 and when the p-values were adjusted using the Benjamini & Hochberg (1995) still had 10865 significantly expressed genes that passed correction. I used the Benjamini & Hochberg (1995) because it was the one seen in lecture and I believe used in a homework assigment, so I was more famliar with it.

Using a heatmap we can see the top genes of interest
```{r message=FALSE, warning=FALSE}
heatmap_matrix <- NormalizedData[,2:ncol(NormalizedData)]
rownames(heatmap_matrix) <- NormalizedData$gene
colnames(heatmap_matrix) <- colnames(NormalizedData[,2:ncol(NormalizedData)])

top_hits <- output_hits$gene[output_hits$P.Value<0.05]
heatmap_matrix_tophits <- t(
scale(t(heatmap_matrix[
    which(rownames(heatmap_matrix) %in% top_hits),])))
if(min(heatmap_matrix_tophits) == 0){
    heatmap_col = circlize::colorRamp2(c( 0, max(heatmap_matrix_tophits)), 
                             c( "white", "red"))
  } else {
    heatmap_col =circlize::colorRamp2(c(min(heatmap_matrix_tophits), 0, max(heatmap_matrix_tophits)), c("blue", "white", "red"))
  }
current_heatmap <- ComplexHeatmap::Heatmap(as.matrix(heatmap_matrix_tophits),
                           cluster_rows = TRUE,
                           cluster_columns = TRUE,
                               show_row_dend = TRUE,
                               show_column_dend = TRUE, 
                               col=heatmap_col,
                               show_column_names = TRUE, 
                               show_row_names = FALSE,
                               show_heatmap_legend = TRUE,
                               )
current_heatmap
```
Unfornataulty there seems to be an error and the map didn't fully render

```{r message=FALSE, warning=FALSE}

```

```{r message=FALSE, warning=FALSE}
kable(NormalizedData[1:5,1:5], type="html")
heatmap_matrix <- NormalizedData[,2:ncol(NormalizedData)]
rownames(heatmap_matrix) <- NormalizedData$genes
colnames(heatmap_matrix) <- colnames(NormalizedData[,2:ncol(NormalizedData)])


```
#Refrences
@article{mcquade_coburn_tu_hasselmann_davtyan_blurton-jones_2018, title={Development and validation of a simplified method to generate human microglia from pluripotent stem cells}, volume={13}, DOI={10.1186/s13024-018-0297-x}, number={1}, journal={Molecular Neurodegeneration}, author={Mcquade, Amanda and Coburn, Morgan and Tu, Christina H. and Hasselmann, Jonathan and Davtyan, Hayk and Blurton-Jones, Mathew}, year={2018}}

@article{edgeR,

    author = {Robinson MD and McCarthy DJ and Smyth GK},

    title = {{edgeR} a Bioconductor package for differential

  expression analysis of digital gene expression data},

    journal = {Bioinformatics},

    year = {2010},

    volume = {26},

    number = {7},

    pages = {139--140},

  }

@article{limma,

    author = {Matthew E Ritchie and Belinda Phipson and Di Wu and Yifang Hu and Charity W Law and Wei Shi and Gordon K Smyth},

    title = {{limma} powers differential expression analyses for {RNA}-sequencing and microarray studies},

    journal = {Molecular Neurodegeneration},

    year = {2018},

    volume = {43},

    number = {7},

    pages = {e47},

    doi = {10.1093/nar/gkv007},

  }



@article{GEOmetadb,

  author = {Yuelin Zhu and Sean Davis and Robert Stephens and Paul S. Meltzer and Yidong Chen},

  title = {GEOmetadb: powerful alternative search engine for the Gene Expression Omnibus.},

  journal = {Bioinformatics (Oxford, England)},

  year = {2008},

  month = {Dec},

  day = {01},

  volume = {24},

  number = {23},

  pages = {2798--2800},

  

issn = {1367-4811},

  doi = {10.1093/bioinformatics/btn520},

  url = {http://www.ncbi.nlm.nih.gov/pubmed/18842599},

  language = {eng},

}



@Manual{BiocManager,

  title = {BiocManager: Access the Bioconductor Project Package Repository},

  author = {Martin Morgan},

  year = {2019},

  note = {R package version 1.30.10},

  url = {https://CRAN.R-project.org/package=BiocManager},

}

@Manual{,
    title = {knitr: A General-Purpose Package for Dynamic Report
      Generation in R},
    author = {Yihui Xie},
    year = {2020},
    note = {R package version 1.28},
    url = {https://yihui.org/knitr/},
  }




@Article{,
      author = {Yuelin Zhu and Sean Davis and Robert Stephens and Paul S. Meltzer and Yidong Chen},
      title = {GEOmetadb: powerful alternative search engine for the Gene Expression Omnibus.},
      journal = {Bioinformatics (Oxford, England)},
      year = {2008},
      month = {Dec},
      day = {01},
      volume = {24},
      number = {23},
      pages = {2798--2800},
      abstract = {The NCBI Gene Expression Omnibus (GEO) represents the largest public repository of microarray data. However, finding data in GEO can be challenging. We have developed GEOmetadb in an attempt to make querying the GEO metadata both easier and more powerful. All GEO metadata records as well as the relationships between them are parsed and stored in a local MySQL database. A powerful, flexible web search interface with several convenient utilities provides query capabilities not available via NCBI tools. In addition, a Bioconductor package, GEOmetadb that utilizes a SQLite export of the entire GEOmetadb database is also available, rendering the entire GEO database accessible with full power of SQL-based queries from within R.},
      issn = {1367-4811},
      doi = {10.1093/bioinformatics/btn520},
      url = {http://www.ncbi.nlm.nih.gov/pubmed/18842599},
      language = {eng},
 }











